### DISCUSSION TITLE
Different behavior for union() and or()

### DISCUSSION THREAD

**Discussion by LucaDiba** [colinhacks/zod#3107] [Q&A] [answered]
## Steps to reproduce
1. Create a function that takes both the Zod schema and an object that should have that schema.
```
function testFunction<TypeValidator extends z.ZodFirstPartySchemaTypes>(
  value: z.infer<TypeValidator>,
  schema: TypeValidator
) {}
```

2. Create two schemas: one with union(), one with or().
3. Call that function with the two schemas and an object that does not conform to the schema (it should throw an error).

## Expected behavior
TS results in an `Argument of type [...] is not assignable to parameter of type [...]` error in both cases.

## Actual behavior
TS results in an `Argument of type [...] is not assignable to parameter of type [...]` error only when using the .or() operator.

## Note
For some reason, this only works when the schema is created directly as a function argument.
If the schema is assigned to a constant and then that const is passed to the function, everything works correctly.
See the sandbox link to see this behavior.

## Sandbox
https://codesandbox.io/p/sandbox/zod-union-or-issue-xj3tv9 (file `src/index.ts`).

**Comment by JacobWeisenburger** [comment, marked as answer]
It works correctly when you use `<Schema extends z.Schema>`

```ts
function testFunction<Schema extends z.Schema> (
    value: z.infer<Schema>,
    schema: Schema,
) { }

testFunction(
    {}, // Argument of type '{}' is not assignable to parameter of type '{ key: "foo"; } | { key: "bar"; }'.
    z.union( [
        z.object( {
            key: z.literal( 'foo' ),
        } ),
        z.object( {
            key: z.literal( 'bar' ),
        } ),
    ] ),
)
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
